File: src/stripe_entitlements/internal_auth.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Protocol

from fastapi import Request

_SCOPE = re.compile(r"^[a-z][a-z0-9:_-]{0,127}$")


class WorkloadAuthenticationError(RuntimeError):
    """An expected, sanitized workload credential rejection."""


class WorkloadAuthorizationError(RuntimeError):
    """An expected, sanitized workload-to-owner authorization rejection."""


@dataclass(frozen=True, slots=True)
class WorkloadPrincipal:
    """Verified service identity supplied by the host's workload authenticator."""

    issuer: str
    subject: str
    scopes: frozenset[str]

    def __post_init__(self) -> None:
        for field, value in (("issuer", self.issuer), ("subject", self.subject)):
            if (
                type(value) is not str
                or not value
                or value != value.strip()
                or len(value.encode("utf-8")) > 512
                or any(not character.isprintable() for character in value)
            ):
                raise ValueError(f"workload principal {field} is invalid")
        if type(self.scopes) is not frozenset:
            raise ValueError("workload principal scopes must be a frozenset")
        if len(self.scopes) > 64 or any(
            type(scope) is not str or _SCOPE.fullmatch(scope) is None for scope in self.scopes
        ):
            raise ValueError("workload principal scopes are invalid")


class WorkloadIdentityAdapter(Protocol):
    """Verify a workload credential, including issuer, expiry and replay policy.

    The adapter is the trust boundary. Implementations must authenticate the complete
    credential and validate algorithm, issuer, audience, expiry, not-before, revocation
    and replay policy before returning a principal; merely decoding a token is
    insufficient.
    """

    async def authenticate(self, request: Request) -> WorkloadPrincipal: ...


class WorkloadOwnerAuthorizer(Protocol):
    """Bind a verified workload to the billable owner selected by a request.

    Implementations normally consult host-owned service/tenant grants. A route scope
    alone is intentionally insufficient authority to select every billing owner.
    """

    async def authorize(
        self,
        principal: WorkloadPrincipal,
        owner_external_ref: str,
        required_scope: str,
    ) -> None: ...


class RejectAllWorkloadIdentityAdapter:
    """Safe default for internal APIs until the host injects workload identity."""

    async def authenticate(self, request: Request) -> WorkloadPrincipal:
        del request
        raise WorkloadAuthenticationError("no workload identity adapter is configured")


class RejectAllWorkloadOwnerAuthorizer:
    """Safe default until the host explicitly binds workloads to billing owners."""

    async def authorize(
        self,
        principal: WorkloadPrincipal,
        owner_external_ref: str,
        required_scope: str,
    ) -> None:
        del principal, owner_external_ref, required_scope
        raise WorkloadAuthorizationError("no workload owner authorizer is configured")

File: typescript/src/internal-auth.ts
const WORKLOAD_SCOPE = /^[a-z][a-z0-9:_-]{0,127}$/u;

export class WorkloadAuthenticationError extends Error {}

export class WorkloadAuthorizationError extends Error {}

function workloadVisible(value: unknown, field: string): string {
  if (
    typeof value !== "string" ||
    value.length === 0 ||
    value !== value.trim() ||
    Buffer.byteLength(value, "utf8") > 512 ||
    Array.from(value).some((character) => /\p{C}/u.test(character))
  ) {
    throw new TypeError(`workload principal ${field} is invalid`);
  }
  return value;
}

class ImmutableScopes implements ReadonlySet<string> {
  readonly #values: Set<string>;

  public constructor(values: Iterable<string>) {
    this.#values = new Set(values);
  }

  public get size(): number {
    return this.#values.size;
  }

  public has(value: string): boolean {
    return this.#values.has(value);
  }

  public forEach(
    callbackfn: (
      value: string,
      value2: string,
      set: ReadonlySet<string>,
    ) => void,
    thisArg?: unknown,
  ): void {
    for (const value of this.#values) {
      callbackfn.call(thisArg, value, value, this);
    }
  }

  public entries(): SetIterator<[string, string]> {
    return this.#values.entries();
  }

  public keys(): SetIterator<string> {
    return this.#values.keys();
  }

  public values(): SetIterator<string> {
    return this.#values.values();
  }

  public [Symbol.iterator](): SetIterator<string> {
    return this.#values[Symbol.iterator]();
  }

  public readonly [Symbol.toStringTag] = "ImmutableScopes";
}

export interface WorkloadPrincipalOptions {
  readonly issuer: string;
  readonly subject: string;
  readonly scopes: ReadonlySet<string>;
}

/** Verified service identity returned only after complete host credential validation. */
export class WorkloadPrincipal {
  public readonly issuer: string;
  public readonly subject: string;
  public readonly scopes: ReadonlySet<string>;

  public constructor(options: WorkloadPrincipalOptions) {
    this.issuer = workloadVisible(options.issuer, "issuer");
    this.subject = workloadVisible(options.subject, "subject");
    if (!(options.scopes instanceof Set)) {
      throw new TypeError("workload principal scopes must be a Set");
    }
    if (
      options.scopes.size > 64 ||
      [...options.scopes].some(
        (scope) => typeof scope !== "string" || !WORKLOAD_SCOPE.test(scope),
      )
    ) {
      throw new TypeError("workload principal scopes are invalid");
    }
    this.scopes = new ImmutableScopes(options.scopes);
    Object.freeze(this);
  }
}

/** Alias kept explicit so host adapters can wrap framework-specific request objects later. */
export type WorkloadRequest = Request;

export interface WorkloadIdentityAdapter {
  /**
   * Authenticate the complete credential, including algorithm, issuer, audience,
   * expiry, not-before, revocation and replay policy. Decoding alone is insufficient.
   */
  authenticate(request: WorkloadRequest): Promise<WorkloadPrincipal>;
}

export interface WorkloadOwnerAuthorizer {
  /** Bind the verified workload and operation to this exact billable owner. */
  authorize(
    principal: WorkloadPrincipal,
    ownerExternalRef: string,
    requiredScope: string,
  ): Promise<void>;
}

/** Safe default until the host injects workload identity verification. */
export class RejectAllWorkloadIdentityAdapter
  implements WorkloadIdentityAdapter
{
  public authenticate(request: WorkloadRequest): Promise<WorkloadPrincipal> {
    void request;
    return Promise.reject(
      new WorkloadAuthenticationError(
        "no workload identity adapter is configured",
      ),
    );
  }
}

/** Safe default until the host binds workloads to exact billing owners. */
export class RejectAllWorkloadOwnerAuthorizer
  implements WorkloadOwnerAuthorizer
{
  public authorize(
    principal: WorkloadPrincipal,
    ownerExternalRef: string,
    requiredScope: string,
  ): Promise<void> {
    void principal;
    void ownerExternalRef;
    void requiredScope;
    return Promise.reject(
      new WorkloadAuthorizationError(
        "no workload owner authorizer is configured",
      ),
    );
  }
}
